### ISSUE TITLE
Chm TOC shows wrong characters for many languages

### ISSUE THREAD

**Issue by lade** [dita-ot/dita-ot#1271]
As one example, the letter Ά shows incorrectly in TOC for chm for Greek. By changing html output to 'windows-1253' instead of 'iso-8859-7' in codepages.xml I got it to show correctly.

The compiler of HTML Help Workshop uses different encodings from what are used by DITA-OT currently. The ones that work correctly are:
- windows-1250 for East European languages (instead of iso-8859-2)
- windows-1252 for Western languages (instead of iso-8859-1)
- windows-1253 for Greek (instead of iso-8859-7)
- windows-1254 for Turkish (instead of iso-8859-9)
- windows-1257 for Estonian (instead of iso-8859-1)

In some of these cases the differences are only about some special characters. And windows-1252 is a superset of iso-8859-1 (except for control characters in the range 80-9F). But in some there are more relevant differences, and it would be nice if the correct codepage would be used.

**Comment by robander** [comment]
This has just come up as an urgent issue for one of my customers.

In the past, we had a hacked version of DITA-OT that converted _some_ characters to entities, but not all, followed by a codepage conversion. What made that work was that the entity conversion was language-aware, so for characters like Ý that exist in iso-8859-2, we do not generate `&Yacute;` (keep the character as Ý) when the target codepage is iso-8859-2. The `convert-lang` utility in DITA-OT always does the conversion, so common letters in iso-8859-2 languages are corrupted. Similarly, I've always ignored entities for Greek characters when the input file is Greek.

If nothing else - could we just add a property that allows skipping the `convert-lang` task, so that when all characters are common/expected characters from the language's CHM charset, everything will work? We've got similar properties on a lot of targets, but none available here. @jelovirt I don't know if adding such a property qualifies as a bug fix for a patch, as it doesn't quite _fix_ a bug but would allow us to work around one...

**Comment by robander** [comment]
FWIW our earlier process (run outside of DITA-OT) was actually based on target codepage rather than language. For example, if target codepage is windows-1250 (as with Czech), none of the common Czech characters are converted. In this case if I'm reading right, we convert the HHP file to 1250 but the HHC contents file, HHK index, and HTML files to the HTML codepage. I'm not sure if that's causing additional trouble -- it makes sense that HTML would be converted to the HTML codepage (with charset properly updated), but I would think that the CHM compiler on Windows would want the other project files (not just HHP) to use the Windows codepage. :/

**Comment by robander** [comment]
Hmm, as a test I've just changed this line that converts chars:
```
                if (entityMap.containsKey(key)) {
```

To this:
```
                if (entityMap.containsKey(key) &&
                        !entityExceptionChar(charset,charCode)) {
```

With the following function (for now just testing Y-acute (Ý and ý):
```
    private boolean entityExceptionChar(final String charset, final int charCode) {
        if (charset.equals(CODEPAGE_ISO_8859_2) &&
                (charCode == 221 || charCode == 253)) 
            return true;
        return false;
    }
```

I appear to get the characters generated / converted properly. I've got a full list of characters that we've been *not* converting in this way for Latin-1, Latin-2, and Greek -- if this approach looks good I could make this as a fix. @jelovirt does that look like a valid approach?

**Comment by robander** [comment]
I'm sure there's a more efficient way of looking these values up, but so I don't lose them, here's the list of characters that we've avoided converting for both 1252 / ISO-8859-1 and 1250 / ISO-8859-2:
```
if (    charCode == 193 || charCode == 225 ||    //A-acute
                    charCode == 194 || charCode == 226 ||    //A-circumflex
                    charCode == 196 || charCode == 228 ||    //A-umlaut
                    charCode == 199 || charCode == 231 ||    //C-cedilla
                    charCode == 201 || charCode == 233 ||    //E-acute
                    charCode == 203 || charCode == 235 ||    //E-umlaut
                    charCode == 205 || charCode == 237 ||    //I-acute
                    charCode == 206 || charCode == 238 ||    //I-circumflex
                    charCode == 212 || charCode == 244 ||    //O-circumflex
                    charCode == 214 || charCode == 246 ||    //O-umlaut
                    charCode == 211 || charCode == 243 ||    //O-acute
                    charCode == 218 || charCode == 250 ||    //U-acute
                    charCode == 220 || charCode == 252 ||    //U-umlaut
                    charCode == 221 || charCode == 253 ||    //Y-acute
                    charCode == 223 || charCode == 215) {   //szlig, times
```

Here are the characters we avoid converting just for 1252 / ISO-8859-1:
```
if (    charCode == 192 || charCode == 224 ||   //A-grave
                    charCode == 195 || charCode == 227 ||    //A-tilde
                    charCode == 197 || charCode == 229 ||    //A-ring
                    charCode == 198 || charCode == 230 ||    //AElig
                    charCode == 200 || charCode == 232 ||    //E-grave
                    charCode == 202 || charCode == 234 ||    //E-circumflex
                    charCode == 204 || charCode == 236 ||    //I-grave
                    charCode == 207 || charCode == 239 ||    //I-uml
                    charCode == 208 || charCode == 240 ||    //ETH
                    charCode == 209 || charCode == 241 ||    //N-tilde
                    charCode == 210 || charCode == 242 ||    //O-grave
                    charCode == 213 || charCode == 245 ||    //O-tilde
                    charCode == 216 || charCode == 248 ||    //O-slash
                    charCode == 217 || charCode == 249 ||    //U-grave
                    charCode == 219 || charCode == 251 ||    //O-circumflex
                    charCode == 222 || charCode == 254 ||    //Thorn
                    charCode == 255) {                       //y-umlaut
```

In addition, for Greek, we avoid converting the full Greek alphabet.

Finally, for Korean, we _do_ convert \ to ₩.

**Comment by robander** [comment]
Have a suggested fix linked above. When converting from the ugly test in my previous comment to use sets, I noticed just one character was missing from the Latin-2 set, which is converted to `&Scaron;` and `&scaron;` (characters 352 and 353).

The fix also switches the HHC and HHK files to use the Windows codepage, just like the HHP does today. Without this change, characters that use different codepoints in ISO-8859-2 versus Windows-1250 are still corrupted (as with the original submitted defect report here).

With the fix above, using my Windows 7 system + DITA-OT 3.0 + the fix in #2851, I'm able to generate English, Czech, and Greek CHM files that compile on my system and display the correct TOC + Index.

- My English sample uses characters that exist in the Latin-1 codepages (like Á) that currently generate symbols (`&Aacute;`) that break the TOC/index. These are no longer converted, so the TOC works. Characters that do not exist in Latin-1 (like Ω) are still converted to entities that do not display, but this is a limitation of the format.
- My Czech sample includes the full Czech alphabet: `Aa Áá Bb Cc Čč Dd Ďď Ee Éé Ěě Ff Gg Hh Ii Íí Jj Kk Ll Mm Nn Ňň Oo Óó Pp Qq Rr Řř Ss Šš Tt Ťť Uu Úú Ůů Vv Ww Xx Yy Ýý Zz Žž`. Many of these are corrupted today, all characters now display properly in TOC and Index.
- Greek sample does the same with alphabet `ΑΒΓΔΕΖΗΘΙΚΛΜΝΞΟΠΡΣΤΥΦΧΨΩ αβγδεζηθικλμνξοπρςστυφχψω`. These characters are still converted to entities for English (works in the content, fails in TOC), but are _not_ converted for Greek (the codepage conversion to 1253 handles them all, and they all work in the TOC).

**Comment by robander** [comment]
Fixed with #2852
